File: go/src/2020/sept/subarray-product-less-than-k.go
package sept

func less(nums []int, k int) int {
	var count = 0
	var product = 1
	for i,j := 0,0;j<len(nums);j++ {
		product *= nums[j]
		for j >= i && product >= k {
			product /= nums[i]
			i++
		}
		count += j-i+1
	}
	return count
}

File: archive/java/src/main/java/cn/az/code/year2020/sept/SubarrayProductLessThanK.java
package cn.az.code.year2020.sept;

/**
 * little-code
 *
 * @author azusachino
 * @since 28/9/2020 21:17
 */
public class SubarrayProductLessThanK {

    public int subArray(int[] nums, int k) {
        if (k == 0) {
            return 0;
        }
        int cnt = 0;
        int pro = 1;
        for (int i = 0, j = 0; j < nums.length; j++) {
            pro *= nums[j];
            // sliding window
            while (i <= j && pro >= k) {
                pro /= nums[i++];
            }
            cnt += j - i + 1;
        }
        return cnt;
    }
}
